### PR TITLE
Javascript Client

### PR DESCRIPTION
**Pull request by baloyan**
File provides error status of various functionalities used in CacheCash
- [ ] cachecash_error.js

Includes all the objects needed by the Client. Also file provides functions, which initialize these objects.
- [ ] client_data.js

An instance of the client requests content from a content provider, but instead of serving the content directly, the content provider returns a ticket bundle, which the client uses to retrieve the content from multiple caches.
- [ ] client.js

A file provides CacheCash project's constant values.
- [ ] constants.js

Includes all the crypto related functions needed in CacheCash.
- [ ] crypto.js

File provides functions in order to create Client object and establish communications between Client - Content Provider (CP) and Client - Caches. (Send a content request, generate ticket bundle, send a ticket lottery)
- [ ] data_service_client.js

File provides CacheCash's message types.
- [ ] message_types.js

File provides data structures, which are needed by Client in order to make a requests from Content Provider and Caches.
- [ ] objects.js

Includes all the utility functions that are used by several modules and not falling under specific module to place them there.
- [ ] utils.js

File provides all the crypto related verification functions.
- [ ] verify.js

### CODE REVIEW

File: javascript_client/client.js [hunk 1]
```diff
@@ -0,0 +1,105 @@
...
+ *   It should be unique, because different clients cannot use the same public
+ *   and private key at the same time. Since we generated them manually and so
+ *   far we were able to generate four examples.
+ *   TODO later it is going to be automated and user is not going to pass any
+ *   argument value, but in some way we have to keep uniqueness of every user.
+ */
+function main() {
+
+  if (process.argv.length != 3) {
+    console.log("Usage: " + __filename + " <client number>");
```
Comment (lukpueh) [inline javascript_client/client.js]: I guess this is 
```javascript
console.log("Usage: node " + __filename + " <client number>");
```
?

File: javascript_client/client.js [hunk 2]
```diff
@@ -0,0 +1,105 @@
...
+      /**
+       * Establish a connection with caches and request content,
+       * compute hash encryption puzzle and generate ticket lottery two.
+       * It also creates a new socket for each connection.
+       */
+      dataservice.RequestFromCaches(client, tkts_bundle, data_block_request, data_blocks);
+    });
+}
+
+main();
```
Comment (lukpueh) [inline javascript_client/client.js]: You should add a `FIXME` that this function call has to be bound to some browser event (e.g.: a button click ?).

File: javascript_client/client.js [hunk 3]
```diff
@@ -0,0 +1,105 @@
...
+    process.exit(1);
+  }
+
+  var client = new client_import.ClientData();
+  client.Init(argv);
+
+  // Tickets bundle sent by the CP as a response to the content request.
+  var tkts_bundle = new objects.TicketsBundle();
+  dataservice.RequestFromContentProvider(client, tkts_bundle,
+    function() {
```
Comment (lukpueh) [inline javascript_client/client.js]: We use two spaces for line continuation.

File: javascript_client/client.js [hunk 4]
```diff
@@ -0,0 +1,105 @@
...
+    cc_error.CacheCashError(cc_error_code.INPUT_PARSING_ERROR,
+                            "Client number must be 1, 2, 3, or 4");
+    process.exit(1);
+  }
+
+  var client = new client_import.ClientData();
+  client.Init(argv);
+
+  // Tickets bundle sent by the CP as a response to the content request.
+  var tkts_bundle = new objects.TicketsBundle();
```
Comment (lukpueh) [inline javascript_client/client.js]: Why do you instantiate a ticket bundle here when you will receive one from the content provider?

File: javascript_client/client.js [hunk 5]
```diff
@@ -0,0 +1,105 @@
...
+  dataservice.RequestFromContentProvider(client, tkts_bundle,
+    function() {
+      /**
+       * Setup the datablocks indices inside the structure that will hold the
+       * retrieved datablocks, note something: we are sticking to the thing that
+       * the ciphertext of a datablock is of an equal size to the plaintext.
+       */
```
Comment (lukpueh) [inline javascript_client/client.js]: I don't find this comment helpful.

Comment (lukpueh) [inline javascript_client/client.js]: I would like to see a comment that explains when this function is executed and what it does.

File: javascript_client/client.js [hunk 6]
```diff
@@ -0,0 +1,105 @@
...
+      data_blocks = new Array(client.blocks);
+
+      /**
+       * Now the client will start communication with the selected set of caches
+       * to retrieve the data blocks.
+       * TODO Basically, the client will create N threads each thread will take care
+       * of completing a full service session with the cache to get the data block.
+       */
```
Comment (lukpueh) [inline javascript_client/client.js]: This does not look like starting a communication. What happens here?

**Comment by lukpueh** [comment]
Consider adding a `package.json` for your front-end dependencies.

File: javascript_client/cachecash_error.js [hunk 7]
```diff
@@ -0,0 +1,68 @@
...
+  MERKLE_TREE_ROOT_NOT_MATCH_ERROR: 77,
+  FILE_FORMAT_ERROR: 80,
+  CONTENT_REQUEST_TRIALS_EXCEEDED_ERROR: 85,
+  CONTENT_REQUEST_ERROR: 86,
+};
+
+
+
+/**
+ * Function prints a meaningful error message beside the error code.
```
Comment (baloyan) [inline javascript_client/cachecash_error.js]: Addressed in https://github.com/baloyan/CacheCash/pull/12

File: javascript_client/cachecash_error.js [hunk 8]
```diff
@@ -0,0 +1,68 @@
...
+ * The printed message has the following format:
+ * error:[error code], descriptive message
```
Comment (baloyan) [inline javascript_client/cachecash_error.js]: Addressed in https://github.com/baloyan/CacheCash/pull/12

File: javascript_client/cachecash_error.js [hunk 9]
```diff
@@ -0,0 +1,68 @@
...
+
+
+/**
+ * Assign any expression in cachecash_error.js that we want to become available
+ * in other files.
+ */
+module.exports = {
+  CACHECASH_ERROR: CACHECASH_ERROR,
+  CacheCashError: CacheCashError
+}
```
Comment (baloyan) [inline javascript_client/cachecash_error.js]: Addressed in https://github.com/baloyan/CacheCash/pull/12

File: javascript_client/client.js [hunk 10]
```diff
@@ -0,0 +1,105 @@
...
+ * Arguments:
+ * - Client number, which is unique integer number for every client. Currently
+ *   you can pass only 1, 2, 3 or 4.
+ *   It should be unique, because different clients cannot use the same public
+ *   and private key at the same time. Since we generated them manually and so
+ *   far we were able to generate four examples.
+ *   TODO later it is going to be automated and user is not going to pass any
+ *   argument value, but in some way we have to keep uniqueness of every user.
+ */
+function main() {
```
Comment (baloyan) [inline javascript_client/client.js]: Addressed in https://github.com/baloyan/CacheCash/pull/12

File: javascript_client/client.js (hunk 1 above)
Comment (baloyan) [inline javascript_client/client.js]: This was removed in https://github.com/baloyan/CacheCash/pull/12

File: javascript_client/client.js [hunk 11]
```diff
@@ -0,0 +1,105 @@
...
+
+  if (process.argv.length != 3) {
+    console.log("Usage: " + __filename + " <client number>");
+    process.exit(1);
+  }
```
Comment (baloyan) [inline javascript_client/client.js]: This was removed in https://github.com/baloyan/CacheCash/pull/12

File: javascript_client/client.js (hunk 4 above)
Comment (baloyan) [inline javascript_client/client.js]: Addressed in https://github.com/baloyan/CacheCash/pull/12

File: javascript_client/client.js (hunk 5 above)
Comment (baloyan) [inline javascript_client/client.js]: Addressed in https://github.com/baloyan/CacheCash/pull/12

File: javascript_client/client.js (hunk 6 above)
Comment (baloyan) [inline javascript_client/client.js]: Addressed in https://github.com/baloyan/CacheCash/pull/12

**Comment by baloyan** [comment]
This PR was updated and now we have browser environment PR https://github.com/baloyan/CacheCash/pull/12

**Pull request closed without merging**